### DISCUSSION TITLE
RP2 SPI oddness with default pins. I'm foxed.

### DISCUSSION THREAD

**Discussion by peterhinch** [micropython/micropython#11074] [RP2040 / Pico]
I have encountered this at least twice but have failed to produce a simple repro. If a hard SPI interface is instantiated with (e.g)
```python
spi = SPI(1)
```
the interface works, in that signals appear on the expected pins, but the application behaves erratically. By contrast this is reliable:
```python
spi = SPI(1, sck = Pin(10, Pin.OUT), mosi=Pin(11, Pin.OUT), miso=Pin(28, Pin.IN))
```
I encountered this with the NRF24l02 - see [code comment](https://github.com/micropython/micropython-lib/blob/c8603192d1d142fdeb7f9578e000c9834669a082/micropython/drivers/radio/nrf24l01/nrf24l01test.py#L26) - and also with a display driver. The display application would run under some circumstances but not others, e.g. worked on USB connection to PC but not when launched on power up by `main.py`. At one point it worked on USB but only if I ran `rshell` first. None of the symptoms made any sense until I remembered the NRF24l01.

The odd thing is that, in the failing state, a scope or LA shows perfect signals on the pins.

The workround is obvious: always specify pins. But it would be good to know what is going on...

**Comment by robert-hh** [comment]
@peterhinch  The code configures the GPIO mode to GPIO_FUNC_SPI, but nothing besides that. Your example with specifying the pins performs additional configuration. Since you see signals at the LA, it would be interesting if the difference is only related to the miso pin. Could you try to run both variants, but only specify miso?

**Comment by peterhinch** [comment]
I re-tested the three variants as follows
 1. `SPI(1)` failed.
 2. `SPI(1, miso=None`) failed.
 3. `SPI(1, miso=Pin(28, Pin.IN))` passed.

Testing was done by power cycling the system: the above behaviour seems consistent.

**Reply by robert-hh** [reply]
The difference between variant 1 and 3 is the interesting one. 3 should work as well, if you just write `SPI(1, miso=Pin(28))`, since Pin.IN is the default. If you can build rp2 firmware, you could try insert a few lines into machine_spi.c at the end of machine_spi_make_new(), line 175 ff. We had a similar problem with UART, where fast consecutive writes to the UART register has a strange result. I do not have a good SPI test vehicle here.

```
        gpio_set_function(self->sck, GPIO_FUNC_SPI);
        __DSB();
        gpio_set_function(self->miso, GPIO_FUNC_SPI);
        __DSB();
        gpio_set_function(self->mosi, GPIO_FUNC_SPI);
        __DSB();
```

**Reply by robert-hh** [reply]
I uploaded a binary with that change to https://github.com/robert-hh/Shared-Stuff/blob/master/firmware_RP2_SPI_DSB.uf2.
On a second thought I'm not overly convinced that it will help.

**Reply by robert-hh** [reply]
@peterhinch  Which board do you use with a default SPI1 MISO at GPIO28? Isn't the default at GPIO8?

**Reply by peterhinch** [reply]
My application is output-only: I don't use MISO. Yet it seems I have to specify it. Options for SPI(1) are GPIO12 or 28. I don't know which is the default, but GPIO12 is used for something else on the display adaptor so I chose 28.

I am using a Pi Pico with [this display adaptor](https://www.waveshare.com/pico-epaper-4.2.htm). As you suggested the following also works:
```python
self.spi = SPI(1, miso=Pin(28))
```
I can build firmware if there's anything you want me to try.

**Reply by robert-hh** [reply]
> My application is output-only: I don't use MISO. 

That is a valuable information. The default pin for MISO at SPI(1) is GPIO8. So please try to use the default by just using SPI(1), but tie GPIO8 to a fixed level, like GND. 

P.S.: I looked into the code: just specifying Pin(x) does not set/change the mode in machine_pin.c. I may have been set before. Some ports do, but not RP2.

**Reply by robert-hh** [reply]
> I am using a Pi Pico with [this display adaptor](https://www.waveshare.com/pico-epaper-4.2.htm).

Looking into may drawers I found a similar Waveshare E-Ink display. Same size, but without the Pico connectors. But it may work and fail as well with your tests. Which test scripts to you use?

**Reply by peterhinch** [reply]
You are right, GPIO8 is another option, but unfortunately it is also used by the adaptor. Hence 28 is the only one I can use. Is it possible that, in the absence of a specific pin, the SPI constructor assigns GPIO 8 or 12? That might account for my problems as there would be a pin conflict. Thinking about this, it would explain a lot of my observations...

Ideally the SPI driver would not configure any pin if `miso=None` were specified. Also (arguably) if `sck` and `mosi` are specified but not `miso`. 

Re my test scripts I am developing a widget for `nano-gui` and updating [this device driver](https://github.com/peterhinch/micropython-nano-gui/blob/master/DRIVERS.md#53-waveshare-400x300-pi-pico-display). I can push a branch of `nano-gui` if you wish.

**Reply by robert-hh** [reply]
If the board does not specify another default pin for MISO, then GPIO8 is used. Unless you specify a different one in the constructor.

**Reply by peterhinch** [reply]
Thank you for your help, I think a pin conflict explains what has been going on.

A couple of general points:
 1. Where are default pins for RP2 interfaces documented?
 2. In my view if the user specifies `miso=None` the constructor should not assign any pin. The intention seems clear. I think this is a bug: the constructor accepts `None` as a value but does not act on it. Do you agree?

**Reply by robert-hh** [reply]
You are right. The mechanism gets only visible if you study the source code. Like many other details it is not documented. And it is different between ports. Some ports like STM32, SAMD, MIMXRT or ESP8266 do not allow to specify pins, for soft_SPI one has to specify all pins, at the nrf port one has to specify either no pins or all pins, and only ESP32 behaves like you suggest.
Another area that needs more consistency.

**Reply by peterhinch** [reply]
What shall we do about the `miso=None` bug? Should I raise an issue or are you in a position to fix it?

**Reply by robert-hh** [reply]
>  Should I raise an issue or are you in a position to fix it?

Raising an issue does not hurt to keep that aspect visible. I can make a PR to change the behavior like you suggest, but the lacks a consent about the intended behavior. You suggestion looks reasonable to me. Implementing it for rp2 is easy. 
For other ports it may be more difficult. And write-only devices are common, and there are even read-only devices. So having something like miso=None or mosi=None even for ports with a fixed pin mapping to SPI is interesting.

**Reply by peterhinch** [reply]
Now done: https://github.com/micropython/micropython/issues/11082

**Reply by rkompass** [reply]
Is this the right opportunity to ask for a possibility to read the logical state of miso?
I encountered the problem while writing a driver for an ADC by TI. The logical state of miso may indicate the presence of new data with some devices.
My wish/request would be to have a spi.miso_value() function. This saves an extra pin with some of the ADCs.
Perhaps not too hard to add to the spi implementation? I would perfer to know the opinion of the experts before I post into the issue.

**Reply by robert-hh** [reply]
One would have to go though every port to tell, if it can be implemented. As name, spi.value() may be sufficient (or spi.miso()).
On some ports like the RP2 it is already possible to read back the state of the Pin even if it is configured e.g. as SPI. You have to configure the MISO pin as Pin object first, to have a handle for the python code. For example:
```
from machine import SPI, Pin
miso_pin = Pin(8, Pin.IN)
spi=SPI(1, miso=miso_pin)  #  uses pin 8 for miso
state = miso_pin.value()
```

**Reply by robert-hh** [reply]
@peterhinch  Hello Peter. Attached is a machine_spi.c for RP2, which allows to specify each SPI signal as None. In that case, a previous Pin setting is not changed. You may test it, if it fits your use case.
[machine_spi.zip](https://github.com/micropython/micropython/files/11019505/machine_spi.zip)

**Reply by robert-hh** [reply]
Looking to some other ports, it seems a little excessive to un-configure SCK and MOSI as well. Most partial SPI applications are output-only. Input only applications are for instance the XPT2046 touch sensor or the HX711 sensor.

**Reply by robert-hh** [reply]
So that is the state of the cross-check:
- **rp2**     Sample implementation done for miso=None
- **esp32** All of mosi, miso and sck can be set to None in the existing state
- **samd**  miso=None is supported already
- **nrf**     Sample implementation done for miso=None; possible for the other signals as well
- **mimxr**t  Possible. There exists a PR #8236, in which the change can easily be added. Only that PR is not maintained and needs quite a bit of work for rebasing.
- **esp8266** Selecting miso, mosi and sck is not supported.
- **stm32**  Selecting miso, mosi and sck is not supported.
- **renesas-ra** Selecting miso, mosi and sck is not supported.
- **SoftSPI**  Implementation for miso=None is possible & straightforward, at a small performance or code size penalty.

**Reply by peterhinch** [reply]
Your patched code works fine with this invocation:
```python
self.spi = SPI(1, sck = Pin(10), mosi=Pin(11))
```
Where a port doesn't support specifying pins I would hope that the constructor would throw an exception if args were passed. If a port doesn't allow specifying `miso=None` an exception should be thrown if it's attempted.

The RP2 behaviour (without your patch) is simply wrong, with an outcome which can be confusing. At one point I suspected a hardware fault.

**Reply by robert-hh** [reply]
For consistency with the previous setting: If you do not specify miso=xxx, the default value is used. I do not know if you mean that with "fine". If miso shall not be used, you have to specifiy `miso=None`.

**Reply by peterhinch** [reply]
I tested `miso=None` successfully. The fact that not specifying `miso` also worked illustrates the intermittent nature of this fault. Pin conflicts are like that, I guess.

The one thing I'm sure of is that specifying an unused pin (GPIO28) _always_ works...

Re other ports, in some cases (STM32) the hardware does not allow alternative pins. I would suggest that the constructor should work as follows
```python
spi = SPI(1)  # OK
# Or
spi = SPI(mosi=Pin("X8"), miso=Pin("X7"), sck=Pin("X6"))  # The only legal pins
```
The following would throw an exception
```python
spi = SPI(mosi=Pin("X8"), miso=None, sck=Pin("X6"))  # Disallowed by hardware
# Or
spi =  SPI(mosi=Pin("X8"), miso=Pin("X7"), sck=Pin("X1"))  # Disallowed by hardware
```

**Reply by robert-hh** [reply]
> The fact that not specifying miso also worked illustrates the intermittent nature of this fault. Pin conflicts are like that, I guess.

The last setting rules. If you instantiate SPI first and then the Pin at GPIO8, it is configured as Pin object.

I do not know yet how to deal with STM or Renesas. ESP8266 could have the single `miso=None` option like MIMXRT. Anything at changing STM32 code is quite a bit of work, since it is quite convoluted, often changed. And the Renesas people just picked the wrong template for their port. They should have chosen rp2 instead of stm32.

**Reply by robert-hh** [reply]
I added a `miso=None` option to ESP8266. But that's it for the moment.

**Comment by robert-hh** [comment]
So that is the state now of the port_spi branch in my copy of the repository:

- **rp2** Sample implementation done for miso=None
- **esp32** All of mosi, miso and sck can be set to None in the existing state
- **samd** miso=None is supported already
- **nrf** Sample implementation done for miso=None; possible for the other signals as well
- **mimxrt** Sample implementation done for miso=None, based on the actual state of the port.
- **esp8266** Support a keyword parameter miso=None, even is selecting miso, mosi and sck is otherwise not supported.
- **SoftSPI** Sample implementation for miso=None.
- **stm32** Selecting miso, mosi and sck is not supported.
- **renesas-ra** Selecting miso, mosi and sck is not supported.

**Comment by scruss** [comment]
fwiw, it doesn't look like pin 28 is the default for SPI(1):

```
>>> spi = SPI(1)
>>> spi
SPI(1, baudrate=992063, polarity=0, phase=0, bits=8, sck=10, mosi=11, miso=8)
```

**Reply by robert-hh** [reply]
That was the intention, because it could be assigned w/o warning. miso was not used for that application, but Pin8 was needed for something else.
